inference/core/workflows/core_steps/common/utils.py

```diff
@@ -48,38 +50,72 @@ def attach_prediction_type_info(
+    predictions_key: str = "predictions",
+    keypoints_key: str = "keypoints",
+    detection_id_key: str = "detection_id",
+    parent_id_key: str = PARENT_ID_KEY,
+) -> List[Dict[str, Union[sv.Detections, Any]]]:
+    converted_predictions: List[Dict[str, Union[sv.Detections, Any]]] = []
+    for p in predictions:
+        converted_prediction = deepcopy(p)
+        detections = sv.Detections.from_inference(p)
+        if any(keypoints_key in d for d in p[predictions_key]):
+            # keypoints arrays may have different length for each detection hence "object" type is used
+            detections[keypoints_key] = np.array(
+                [
+                    (
+                        np.array(
+                            [
+                                [keypoint["x"], keypoint["y"]]
+                                for keypoint in d[keypoints_key]
+                            ]
+                        )
+                        if d.get(keypoints_key)
+                        else np.array([])
+                    )
+                    for d in p[predictions_key]
+                ],
+                dtype="object",
+            )
+        if any(parent_id_key in d for d in p[predictions_key]):
+            detections[parent_id_key] = np.array(
+                [
+                    d[parent_id_key] if d.get(parent_id_key) else None
+                    for d in p[predictions_key]
+                ]
+            )
+        detection_ids = [
+            d[detection_id_key] if detection_id_key in d else str(uuid.uuid4)
+            for d in p[predictions_key]
+        ]
+        detections[detection_id_key] = np.array(detection_ids)
+        converted_prediction[predictions_key] = detections
+        converted_predictions.append(converted_prediction)
+    return converted_predictions
 
 
-def attach_parent_info_to_prediction(
-    image: Dict[str, Any],
-    prediction: Dict[str, Any],
-    nested_key: Optional[str],
-) -> Dict[str, Any]:
-    prediction[PARENT_ID_KEY] = image[PARENT_ID_KEY]
-    if nested_key is None:
-        return prediction
-    for detection in prediction[nested_key]:
-        detection[PARENT_ID_KEY] = image[PARENT_ID_KEY]
-    return prediction
+def attach_parent_info(
+    images: List[Dict[str, Any]],
+    predictions: List[Dict[str, Union[sv.Detections, Any]]],
+    nested_key: Optional[str] = "predictions",
+) -> List[Dict[str, Union[sv.Detections, Any]]]:
+    for image, prediction in zip(images, predictions):
+        prediction[PARENT_ID_KEY] = image[PARENT_ID_KEY]
+        if nested_key is None:
+            continue
+        detections = prediction[nested_key]
+        detections[PARENT_ID_KEY] = [image[PARENT_ID_KEY]] * len(detections)
+    return predictions
 
 
 def anchor_prediction_detections_in_parent_coordinates(
     image: List[Dict[str, Any]],
-    predictions: List[Dict[str, Any]],
+    predictions: List[Dict[str, Union[sv.Detections, Any]]],
     image_metadata_key: str = "image",
```

Comment:
[question] Is `image_metadata` still needed?

Reply:
I left this here for backwards compatibility, we can consider removing it if it's not required.

Unfortunately `sv.Detections` does not keep image metadata information (`width`/`height`) so I would imagine this metadata needs to be passed to blocks in some way (we could consider moving it into `sv.Detections.data` though it will be the same `width`/`height` copied for each detections, maybe it's not such a big deal since there won't be that many detections, however we will have pattern similar to this: `image_metadata = detections[image_metadata][0]` when we will want to obtain metadata)

Reply:
yeah, if it's truly needed we should probably try to pack it into detections, but the sole purpose of image metadata was to make it possible to draw things out of `workflows` output

Reply:
so if there is no block operating on that property (I guess blocks just pass it by) - we can remove